File: leet_283/leet_283b.cpp
class Solution {
public:
    void moveZeroes(vector<int>& nums) {
        reverse(nums.begin(),nums.end());
        stable_sort(nums.begin(), nums.end(), [](int a, int b){return a!=0 || b==0;});
    }
};
// same speed as first implementation but more concise code

File: leet_easy_java/leet_283/leet_283b.java
class Solution {
    public void moveZeroes(int[] nums) {
        int tmp;
        for(int i=0, j=0; i<nums.length; i++){
            if(nums[i]!=0){
                tmp = nums[j];
                nums[j] = nums[i];
                nums[i] = tmp;
                j++;
            }
        }
    }
}
/*
[#$0, 1, 0, 3, 12]
[#0, $1, 0, 3, 12]
[1, #0, $0, 3, 12]
[1, #0, 0, $3, 12]
[1, 3, #0, 0, $12]
[1, 3, #0, 0, $12]
[1, 3, 12, #0, 0] $
*/